### ISSUE TITLE
TeX Live package updates not trigger cache update

### ISSUE THREAD

**Issue by muzimuzhi** [teatimeguest/setup-texlive-action#329]
See for example this [job log](https://github.com/muzimuzhi/tabularray-dev/actions/runs/11995763618/job/33439409149),

- The installation step updated 26 TeX Live packages.
  ```yaml
  # snippet of workflow file
    - name: Setup TeX Live
      uses: teatimeguest/setup-texlive-action@v3
      with:
        package-file: .github/tl_packages
        update-all-packages: true
   ```
- But the post installation step didn't save a new cache, and only wrote to log:
  ```
  Post job cleanup.
  Cache hit occurred on the primary key setup-texlive-action-linux-x64-2024-..., not saving cache
  ```

**Comment by muzimuzhi** [comment]
The latest `setup-texlive-action` cache in muzimuzhi/tabularray-dev was created on 2024-11-15, although I had pushed several times in the past week.

```bash
$ gh cache list -R muzimuzhi/tabularray-dev -k setup-texlive --json id,key,createdAt,lastAccessedAt,ref
```
(gh cli doc: https://cli.github.com/manual/gh_cache_list)

```json
[
  {
    "createdAt": "2024-11-01T07:31:50.57Z",
    "id": 24,
    "key": "setup-texlive-action-linux-x64-2024-ffb5fc6d4f09b275a355babb70eb154b50c2898e952425a319830934e9ecb971",
    "lastAccessedAt": "2024-11-24T11:33:44.16Z",
    "ref": "refs/heads/develop"
  },
  {
    "createdAt": "2024-11-15T06:27:31.73Z",
    "id": 36,
    "key": "setup-texlive-action-linux-x64-2024-4f53cda18c2baa0c0354bb5f9a3ecbe5ed12ab4d8e11ba873c2f11161202b945",
    "lastAccessedAt": "2024-11-21T09:38:41.4Z",
    "ref": "refs/heads/develop"
  }
]
```

**Comment by muzimuzhi** [comment]
[This job](https://github.com/muzimuzhi/tabularray-dev/actions/runs/11419289257/job/32372048935) used v3.3.2. After updating 2 packages, the installation step ended with log
```
After the job completes, /home/runner/work/_temp/setup-texlive-action/2024 will be saved to cache with key: setup-texlive-action-linux-x64-2024-...
```
then in post-process step it tried to save a new cache.

Therefore it seems it's a regression since [v3.3.3](https://github.com/teatimeguest/setup-texlive-action/releases/tag/v3.3.3), most probably caused by
66d427e (perf: reduce overhead by combining `tlmgr update` calls into one (#<span></span>226), 2024-11-04).

Also, using v3.3.2 the log of installation was consisted of
```
▶ Run teatimeguest/setup-texlive-action@3
...
▶ Restoring cache
▶ Updating tlmgr
▶ Updating packages
▶ Installing packages
▶ TeX Live environment details
```
while using v3.3.3 the sub-steps "▶ Updating tlmgr" and "▶ Installing packages" were missing.

**Comment by teatimeguest** [comment]
Thanks for the suggestion. To make it short, this is the current intended behavior, but I feel the need to revisit it.

---

Some background:

- Basically, the cache key of an installation takes the following form:
  ```
  setup-texlive-action-<os>-<arch>-<texlive version>-<hash>
  ```
  where `<hash>` is computed from the package *names* given by the `packages` and `package-file` inputs. So, the *versions* of the packages do not affect the cache key, and consequently, new cache entries are not created by package updates.

  This seemed to me like a natural choice, since TeX Live does not have a lockfile mechanism and for most users it should not matter which version of a package is installed.

- Another reason for the behavior is that Actions caches are immutable and cannot be updated; however, the action will actually update cache in some cases by using [this workaround](https://github.com/actions/cache/blob/v4.1.2/tips-and-workarounds.md#update-a-cache), which is necessary to handle new TeX Live version releases (#248).

  But at least as of #227, it seemed somewhat hacky and made me hesitant to use it.

After reconsidering the `update-all-packages` input, I think it is more appropriate to update cache when package updates actually happen, as you suggest, by using the workaround.

**Comment by teatimeguest** [comment]
Also, perhaps a shorter name (e.g., just `update`) could be used for the input, as suggested by @pablgonz (https://github.com/teatimeguest/setup-texlive-action/issues/226#issuecomment-1192091356).

**Comment by teatimeguest** [comment]
> while using v3.3.3 the sub-steps "▶ Updating tlmgr" and "▶ Installing packages" were missing.

This is intended:
- Since `tlmgr` is now updated together with other packages with a single call, "▶ Updating tlmgr" is never displayed after v3.3.3.
- When a cache was hit, the action will not attempt to install packages, and so "▶ Installing packages" will not be displayed.

**Comment by muzimuzhi** [comment]
> * Basically, the cache key of an installation takes the following form:
>   ```
>   setup-texlive-action-<os>-<arch>-<texlive version>-<hash>
>   ```
>     
>   where `<hash>` is computed from the package _names_ given by the `packages` and `package-file` inputs.

> * When a cache was hit, the action will not attempt to install packages, and so "▶ Installing packages" will not be displayed.

This also seems not ideal. As scheme and collection names are both accepted, when a new package is added to a scheme or collection, then it will never be installed when cache is hit?

**Comment by muzimuzhi** [comment]
> * This seemed to me like a natural choice, since TeX Live does not have a lockfile mechanism

The local texlive database is stored in `$(kpsewhich --var-value TEXMFROOT)/tlpkg/texlive.tlpdb` and the package info can be queried using [`tlmgr info`](https://www.tug.org/texlive/doc/tlmgr.html#info). (There's also a `tlmgr dump-tlpdb` to dump all, package info, options, etc.)

```
$ tlmgr list --only-installed --data name,localrev | head -5

abraces,71955
accents,51497
accsupp,53052
acro,62925
acronym,72324
```

`tlmgr info` also support output in 'JSON' format (with with trailing commas), with more info:
```
# '.[:1]' is used to show only the first item
$ tlmgr info --only-installed --json | \
jq '.[:1] | [.[] | {name, lrev, rrev, containerchecksum}]'
```

```json
[
  {
    "name": "abraces",
    "lrev": 71955,
    "rrev": 71955,
    "containerchecksum": "b32204970bb51dbe2fd16c49e0a8c9957c1c4ee45ec5f90dd19e70f85d9a924add4042d61a9f054a797ce966b1fe953369fae6eb04d2eaf1c68941945e0e2253"
  }
]
```

It's said the format of JSON output is documented in `tlpkg/doc/json-formats.txt`, format definition: `TLPOBJINFO`, but somehow I don't see it locally, perhaps caused by my local setting.

**Comment by muzimuzhi** [comment]
> It's said the format of JSON output is documented in `tlpkg/doc/json-formats.txt`, format definition: `TLPOBJINFO`, but somehow I don't see it locally, perhaps caused by my local setting. (Asked in the chat room of TeX-SX, see [here](https://chat.stackexchange.com/transcript/message/66681040#66681040).)

Thanks to @<span></span>mbert on TeX-SX (in [this chat message](https://chat.stackexchange.com/transcript/message/66681117#66681117)), the `tlpkg/doc/json-formats.txt` can be found in TeX Live svn repository: https://tug.org/svn/texlive/trunk/Master/tlpkg/doc/json-formats.txt.

**Comment by teatimeguest** [comment]
> This also seems not ideal. As scheme and collection names are both accepted, when a new package is added to a scheme or collection, then it will never be installed when cache is hit? (Or the schemes and collections are already expanded to list of packages when calculating the hash?)

This is in fact true, and schemas and collections are not expanded.

If the action were to do so, the `texlive.tlpdb` file would have to be fetched each time prior to cache restoration, which would be a bit expensive due to its large file size. I do not know how often packages are added to a schema or collection, but I suspect not very often. Therefore, I am concerned that it would be an undesirable and overly cautious behavior for most use cases.

**Comment by teatimeguest** [comment]
> It's said the format of JSON output is documented in tlpkg/doc/json-formats.txt, format definition: TLPOBJINFO, but somehow I don't see it locally, perhaps caused by my local setting. (Asked in the chat room of TeX-SX, see [here](https://chat.stackexchange.com/transcript/message/66681040#66681040).)

Actually, an unofficial documentation can be found [here](https://github.com/teatimeguest/texlive-json-schemas).

Nevertheless, the `tlmgr info` command is quite slow, so the action parses `tlpdb` on its own.

**Comment by muzimuzhi** [comment]
> Therefore, I am not so sure how much sense for users it would make to manage caches based on such strict version information.

There's another sort of request.

Because there's no lock file in LaTeX community and it's very hard to install specific version of a LaTeX package, usually LaTeX package/class maintainers want to test his/her package with all other packages updated to the latest versions.

So if package updates don't trigger cache update/saving, the installation step will get slower and slower, gradually offsetting the time saved by caching.

One of the main cause of slowing down is when package updates trigger format rebuilding. Nowadays the LaTeX format depends on more separate packages (latex itself, l3kernel, l3backend, babel and hyphenation related ones, firstaid, and perhaps more) thus its rebuilding is triggered more often.

For example in the [job](https://github.com/muzimuzhi/tabularray-dev/actions/runs/11995763618/job/33439409149) (linked in my initial https://github.com/teatimeguest/setup-texlive-action/issues/329#issue-2687636779), 10 formats are rebuilt after package updates (pdflatex, xelatex, lualatex, luadvilatex, latex, and 5 corresponding dev formats).

**Comment by teatimeguest** [comment]
Thank you for patiently explaining it to me. I now understand your concern and the importance of stricter cache management to package/class maintainers. I am considering introducing new input `strict`, which will allow cache keys to be computed taking revisions into account, with schemas and collections expanded. This will take a little time as it needs some work.

> After reconsidering the update-all-packages input, I think it is more appropriate to update cache when package updates actually happen, as you suggest, by using the workaround.

I am going to do this first.

**Comment by muzimuzhi** [comment]
Huge thanks!

**Comment by teatimeguest** [comment]
> Also, perhaps a shorter name (e.g., just update) could be used for the input, as suggested by @pablgonz (https://github.com/teatimeguest/setup-texlive-action/issues/226#issuecomment-1192091356).

As for this, I have created a branch [here](https://github.com/teatimeguest/setup-texlive-action/tree/rename-inputs). I am planning to rename inputs (and output) as follows
```diff
- package-file
+ packages-files

- update-all-packages
+ update

- version
+ texlive-version
```
The old names will not be deprecated immediately but will be removed from the README.

I will work on this later.

**Comment by teatimeguest** [comment]
> I am considering introducing new input strict, which will allow cache keys to be computed taking revisions into account, with schemas and collections expanded.

As for this, I have created a new issue (#330).

**Comment by muzimuzhi** [comment]
> In [v3.3.4](https://github.com/teatimeguest/setup-texlive-action/releases/tag/v3.3.4), cache is also updated when a package is updated. This is also the case when `tlmgr` itself is updated with `update-all-packages: false`.

It seems in v3.3.4, new cache is always created. In other words, the cache key is too sensitive to changes which are almost transparent to users.

For example, for the two job runs ([1] and [2]) defined in the same workflow, they shared the same "▶ TeX Live environment details" log (used the same CTAN mirror, had the same `tlmgr` and package versions), but generated different `<hash>` part of cache keys, thus each created their own cache.

```
▶ TeX Live environment details
  /home/runner/work/_temp/setup-texlive-action/2024/bin/x86_64-linux/tlmgr version
  tlmgr revision 71331 (2024-05-24 09:30:36 +0200)
```
(I wrapped the long log line.)
```diff
 After the job completes, /home/runner/work/_temp/setup-texlive-action/2024
   will be saved to cache with key:
   setup-texlive-action-linux-x64-2024-
   ffb5fc6d4f09b275a355babb70eb154b50c2898e952425a319830934e9ecb971-
-  b1894d572bc54a8db3d65d60d12c3b7d
+  58be16a7aef5473bbd5d09be489d4fc9
```

[1]: https://github.com/muzimuzhi/tabularray-dev/actions/runs/12167737083/job/33936963513
[2]: https://github.com/muzimuzhi/tabularray-dev/actions/runs/12167737083/job/33936963848

**Comment by muzimuzhi** [comment]
> > In [v3.3.4](https://github.com/teatimeguest/setup-texlive-action/releases/tag/v3.3.4), cache is also updated when a package is updated. This is also the case when `tlmgr` itself is updated with `update-all-packages: false`.
> 
> It seems in v3.3.4, new cache is always created. In other words, the cache key is too sensitive to changes which are almost transparent to users.

Don't know why but it's not reproducible now, with [teatimeguest/setup-texlive-action@v3.4.1](https://github.com/teatimeguest/setup-texlive-action/releases/tag/v3.4.1).

**Comment by teatimeguest** [comment]
@muzimuzhi Sorry for the late reply. 

> It seems in v3.3.4, new cache is always created. In other words, the cache key is too sensitive to changes which are almost transparent to users.

I think this is due to concurrent jobs, in which case the behavior is expected. Here is why:

GitHub Actions caches are immutable and cannot currently be overwritten with the same key. To work around this limitation, the action [appends a random string to the cache key](https://github.com/actions/cache/blob/v4.2.3/tips-and-workarounds.md#update-a-cache) when saving a new entry to emulate cache updates. So, if:

1. jobs run in parallel,
2. a certain TeX Live cache is restored in several jobs with `cache-hit=true`, and
3. TeX package updates actually occur,

then multiple new cache entries will be created by separate keys.

On the other hand, if any of these conditions are not met, it does not happen:

1. Obvious;
2. When `cache-hit=false`, the cache key does not have a random suffix (i.e., it is deterministic), preventing multiple entries from being created;
    - (though, I am now realizing that this might be considered inconsistent behavior)
3. No new cache registration will be attempted.

Therefore,

> Don't know why but it's not reproducible now, with [teatimeguest/setup-texlive-action@v3.4.1](https://github.com/teatimeguest/setup-texlive-action/releases/tag/v3.4.1).

I would think this is not a version-specific issue, but rather depends on some other conditions. What do you think?

**Comment by teatimeguest** [comment]
I think a workaround is to separate out the cache creation job using the `needs` condition (https://github.com/teatimeguest/setup-texlive-action/issues/317#issuecomment-2395217420).

**Comment by muzimuzhi** [comment]
> I would think this is not a version-specific issue

Agreed.

I cannot recall full details, especially when the run logs (linked in https://github.com/teatimeguest/setup-texlive-action/issues/329#issuecomment-2518562909) have expired now. It seems the problem was that even when each of the parallel job runs created their own cache, rerun of the same job still created new caches.

I will come back with attached full job run logs next time.

> I think a workaround is to separate out the cache creation job using the `needs` condition ([#317 (comment)](https://github.com/teatimeguest/setup-texlive-action/issues/317#issuecomment-2395217420)).

Even with the default `update-all-packages: false` input value, the parallel jobs may still trigger cache update, if `tlmgr` happens to be updated, do they?

```yaml
jobs:
  ensure-cached:
    runs-on: ubuntu-latest
    steps:
      - uses: teatimeguest/setup-texlive-action@v3
        with:
          package-file: .github/tl_packages
          update-all-packages: true
  ci:
    needs: ensure-cached
    runs-on: ubuntu-latest
    steps:
      - uses: teatimeguest/setup-texlive-action@v3
        with:
          package-file: .github/tl_packages
          update-all-packages: false # default value
      - ...
```

> Instead of a random string, the [`GITHUB_RUN_ID` environment variable](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables) might be used for cache keys.

Along with `GITHUB_RUN_ATTEMPT`, as `GITHUB_RUN_ID` doesn't take run attempts into account.

**Comment by teatimeguest** [comment]
> It seems the problem was that even when each of the parallel job runs created their own cache, rerun of the same job still created new caches.

I am not sure if re-runs would limit cache functionality, but which events trigger a workflow will certainly affect [cache visibility](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache).

> Even with the default update-all-packages: false input value, the parallel jobs may still trigger cache update, if tlmgr happens to be updated, do they?

Yes. If a new version of `tlmgr` is released between

- the time the cache key is computed in the `ensure-cached` job and
- the time the cache is restored in the `ci` job,

then new cache entries are created at the end of the `ci` job, regardless of the value of `update-all-packages`. Currently the action does not provide a way to prevent this, but I expect it to happen rarely.

Or, it might be useful to allow fine control over cache. For example, how about extending the `cache` input with two new values, `restore-only` and `save-only`? Setting `cache: restore-only` in the `ci` job would suppress cache creation.

> > Instead of a random string, the [GITHUB_RUN_ID environment variable](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables) might be used for cache keys.
>
> Along with GITHUB_RUN_ATTEMPT, as GITHUB_RUN_ID doesn't take run attempts into account.

That's right. I hope to work on this in the next few days.

**Comment by muzimuzhi** [comment]
To be honest, recently I haven't been doing much on repos which uses `teatimeguest/setup-texlive-action`, so no hurry at all.
